autonat/autonat-v2.md

```diff
@@ -0,0 +1,193 @@
+peers to establish a relayed connection.
+
+In `autonat v2` client sends a priority ordered list of addresses. On receiving
+this list the server dials the first address on the list that it is capable of
+dialing. `autonat v2` allows nodes to determine reachability for individual
+addresses. Using `autonat v2` nodes can build an address pipeline where they can
+test individual addresses discovered by different sources like identify, upnp
+mappings, circuit addresses etc for reachability. Having a priority ordered list
+of addresses provides the ability to verify low priority addresses.
+Implementations can generate low priority address guesses and add them to
+requests for high priority addresses as a nice to have. This is especially
+helpful when introducing a new transport. Initially, such a transport will not
+be widely supported in the network. Requests for verifying such addresses can be
+reused to get information about other addresses
+
+Compared to `autonat v1` there are two major differences
+1. `autonat v1` allowed testing reachability for the node. `autonat v2` allows
+testing reachability for an individual address
+2. `autonat v2` provides a mechanism for nodes to verify whether the peer
+actually successfully dialled an address.
+
+
+## AutoNAT V2 Protocol
+
+A node wishing to determine reachability of its adddresses sends a `DialRequest`
+message to a peer on a stream with protocol ID
+`/libp2p/autonat/2.0.0/dial-request`. 
+
+This `DialRequest` message has a list of `AddressDialRequest`s. Each item in
+this list contains an address and a fixed64 nonce. The list is ordered in
+descending order of priority for verfication.
+
+Upon receiving this message the peer attempts to dial the first address from the
+list that it is capable of dialing. It dials this address, opens a stream with
+Protocol ID `/libp2p/autonat/2.0.0/dial-attempt` and sends a `DialAttempt`
+message with the nonce received in the corresponding `AddressDialRequest`. The
+peer MUST NOT dial any address other than the first address in the list that it
+is capable of dialing.
+
+Upon completion of the dial attempt, the peer sends a `DialResponse` message to
+the initiator node on the `/libp2p/autonat/2.0.0/dial-request` stream with the
+index(0 based) of the address that it attempted to dial and the appropriate
+`ResponseStatus`. see [Requirements For
+ResponseStatus](#requirements-for-responsestatus)
+
+The initiator MUST check that the nonce received in the `DialAttempt` is the
+same as the nonce the initiator sent in the `AddressDialRequest` for the address
+index received in `DialResponse`. If the nonce is different, the initiator MUST
+discard this response.
+
+
+### Requirements for ResponseStatus
+
+On receiving a `DialRequest` the peer selects the first address on the list it
+is capable of dialing. This address is referred to as _addr_. The
+`ResponseStatus` sent by the peer in the `DialResponse` message MUST be set
+according to the following requirements
```

Comment:
Instead of implicitly dialing only the first one we support, have we considered responding with a status code for each one by sending one `DialResponse` per address (on the same stream)?

This would provide a lot more information to a node about what addresses are supported and we can for example send the once a node can't parse or dial to a different autonat server.

Reply:
autonat v2 started with the idea of sending a single address in a message. This is fine because creating a stream is cheaper than dialing an address, so the initiator can just open a new stream for all addresses it wants verified. 

While running a node, we may discover addresses or guess at our addresses(quic-v1 => webtransport). Depending on our node state we may or may not want to verify these addresses. If we already have a reachable tcp address, we probably don't want to verify another tcp address. On the other hand some addresses we do want to verify like a webrtc whose reachability we don't know. Requests to nodes for dialing webrtc addresses will likely fail because the support in the network is low. In this case we can add the lower priority tcp address to the list as a "nice to have". If there was no higher priority webrtc address, we'd not have tried verifying this low priority tcp address. 

The expectation is that nodes are not sending every address they know in every request. 
This previous discussion is [here](https://github.com/libp2p/specs/issues/539) 

@marten-seemann, thoughts? 


@thomaseizinger I'd love to get your thoughts on this

Reply:
Nice idea. As I understand browser nodes will have a lot of such cases where they understand the address but cannot dial it. 

I don't see any reason why not to do this. It only provides more information. 

@marten-seemann @mxinden any thoughts?

Reply:
I have incorporated this suggestion.